### PR TITLE
Text Classification LSTM Livebook & Example Integration

### PR DESCRIPTION
**Pull request by fantypants**
Hello, 

This is the initial implementation of the LSTM NER Text Classification. I've worked on the live-book & example for the model.

### CODE REVIEW

File: examples/classifcation/text_classification.exs [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+# Test Data: https://github.com/bhuvanakundumani/NER_tensorflow2.2.0/blob/master/data/test.txt
+
+# Train Data: https://github.com/bhuvanakundumani/NER_tensorflow2.2.0/blob/master/data/train.txt
+
+# Validation Data: https://github.com/bhuvanakundumani/NER_tensorflow2.2.0/blob/master/data/valid.txt
+
+Mix.install([
+  {:axon, github: "elixir-nx/axon"},
+  {:nx, "~> 0.2.1"},
+  {:exla, "~> 0.2.2"}
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:15]: If using Axon GH you should also use these git deps:

```suggestion
  {:exla, "~> 0.3.0-dev", github: "elixir-nx/nx", sparse: "exla"},
  {:nx, "~> 0.3.0-dev", github: "elixir-nx/nx", sparse: "nx", override: true},
```

File: examples/classifcation/text_classification.exs [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+  @batch_size 128
+  @lstm 128
+  @embed_dimension 256
+  @def_split " "
+  @start "-DOCSTART- -X- -X- O\n"
+  @ending "-------------------------------------------------------------"
+  @incorrect [".", "?", "%", ",", ":", ";", "-docstart-"]
+
+  # seqlen
+  # word_count
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:35]: Guessing just forgot to remove?

```suggestion
```

File: examples/classifcation/text_classification.exs [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+  def train(model, data) do
+    model
+    |> Axon.Loop.trainer(&kl_divergence/2, Axon.Optimizers.adam(0.001))
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:39]: You should be able to just use the atom here:

```suggestion
    |> Axon.Loop.trainer(:kl_divergence, Axon.Optimizers.adam(0.001))
```

File: examples/classifcation/text_classification.exs [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+    |> Axon.Loop.metric(:accuracy, "Accuracy")
+    |> Axon.Loop.run(data, %{}, epochs: 3, iterations: 5)
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:41]: Only 5 iterations?

File: examples/classifcation/text_classification.exs [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+  end
+
+  defp kl_divergence(y_true, y_pred) do
+    Axon.Losses.kl_divergence(y_true, y_pred, reduction: :mean)
+  end
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:47]: You can remove once you replace with an atom `:kl_divergence` above
```suggestion
```

File: examples/classifcation/text_classification.exs [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+      Axon.input({nil, @sequence_length, 1}, "inputs")
+       |> Axon.embedding(word_count, @sequence_length)
+       |> Axon.nx(fn t ->
+         t[[0..-1//1, -1]] # Dropping out the sized 1 column created from the embed
+       end)
+       |> Axon.spatial_dropout(rate: 0.1)
+       |> Axon.lstm(@lstm)
+       |> then(fn {{new_cell, new_hidden}, out} ->
+         out
+       end)
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:59]:
```suggestion
    |> elem(1)
```

File: examples/classifcation/text_classification.exs [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+    |> Axon.Loop.run(data, %{}, epochs: 3, iterations: 5)
+
+  end
+
+  defp kl_divergence(y_true, y_pred) do
+    Axon.Losses.kl_divergence(y_true, y_pred, reduction: :mean)
+  end
+
+  def build_model(word_count, label_count) do
+      Axon.input({nil, @sequence_length, 1}, "inputs")
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:50]: New Axon doesn't require input shapes :)

You can still give it a hint though

```suggestion
      Axon.input("inputs", shape: {nil, @sequence_length, 1})
```

File: examples/classifcation/text_classification.exs [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+       |> Axon.embedding(word_count, @sequence_length)
+       |> Axon.nx(fn t ->
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:52]: The reason you have sized 1 is because your input shape is `{nil, seq_len, 1}`. My suggestion is to squeeze your inputs such that the new input shape is `{nil, seq_len}` instead, and then you can remove this

File: examples/classifcation/text_classification.exs [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+         t[[0..-1//1, -1]] # Dropping out the sized 1 column created from the embed
+       end)
+       |> Axon.spatial_dropout(rate: 0.1)
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:55]: I'm not sure this is a good idea---a better approach would be to use recurrent dropout, but Axon does not support that yet :)

File: examples/classifcation/text_classification.exs [hunk 10]
```diff
@@ -0,0 +1,273 @@
...
+    train_data =
+      word_labels
+      |> Enum.map(fn {word, label} ->
+        encode(word_to_idx, word)
+      end)
+      |> Enum.chunk_every(@sequence_length, 1, :discard)
+      |> Enum.drop(-1)
+      |> Nx.tensor
+      |> Nx.divide(wcount)
+      |> Nx.reshape({:auto, @sequence_length, 1})
```
Comment (seanmor5) [inline examples/classifcation/text_classification.exs:92]: Relates to shape comment above :)

```suggestion
      |> Nx.reshape({:auto, @sequence_length})
```

**Review by seanmor5 — COMMENTED** [review]
Nice work! I've left some comments. Also, I recommend running the formatter on the script, I think there are some minor formatting things :)

File: examples/classifcation/text_classification.exs (hunk 9 above)
Comment (fantypants) [reply examples/classifcation/text_classification.exs:55]: I like this feedback a lot, using spatial_dropout results in some strange things happening, I've been doing testing on this. So you propose recurrent dropout? I'll have to look into what it is and see what we can do with it. The spatial dropout I believe seems to cause issues after X epochs, I assume it's due to it dropping same channel across the tensors, which could be the feature map.

File: examples/classifcation/text_classification.exs (hunk 6 above)
Comment (fantypants) [reply examples/classifcation/text_classification.exs:59]: This is why we learn from you haha

File: examples/classifcation/text_classification.exs (hunk 5 above)
Comment (fantypants) [reply examples/classifcation/text_classification.exs:47]: perfect!

File: examples/classifcation/text_classification.exs (hunk 4 above)
Comment (fantypants) [reply examples/classifcation/text_classification.exs:41]: Oversight from early testing, I've been running at 100,300,500,600,800 iterations at current, way back at the start i wanted to see what was actually happening with the model loading & saving etc, was just easy for testing. This is something I need to also cover a bit more in the live book as it's an important key factor to understand I believe (I'm completely self taught other than what I've learnt from yourself, recent mentoring in the space). I'll need to do some more research into how elaborate on the hyper-parameters as a whole

File: examples/classifcation/text_classification.exs (hunk 10 above)
Comment (fantypants) [reply examples/classifcation/text_classification.exs:92]: So this was a weird bit for me, I followed alot of tutorials at the start for the basic concept, and the sizing were always not well explained. I'll do as suggested above, as it makes more sense, I believe the problem at the start I was having was due to the embedding layer and me not understanding I needed to squeeze it out earlier.

File: examples/classifcation/text_classification.exs [hunk 11]
```diff
@@ -0,0 +1,273 @@
...
+  end
+
+  defp encode(dictionary, word) do
+    with {:ok, encoded} <- Map.fetch(dictionary, word) do
+      encoded
+    else
+      _->
+      0
+    end
+  end
```
Comment (jonatanklosko) [inline examples/classifcation/text_classification.exs:71]:
```suggestion
  defp encode(dictionary, word), do: dictionary[word] || 0
```

Sidenote: instead of with like this, you can use `case` :)

File: examples/classifcation/text_classification.exs [hunk 12]
```diff
@@ -0,0 +1,273 @@
...
+
+  defp decode(dictionary, word) do
+    with {:ok, decoded} <- Map.fetch(dictionary, word) do
+      decoded
+    else
+      _-> "Unknown"
+    end
+  end
```
Comment (jonatanklosko) [inline examples/classifcation/text_classification.exs:79]:
```suggestion
  defp decode(dictionary, word), do: dictionary[word] || "Unknown"
```

File: examples/classifcation/text_classification.exs [hunk 13]
```diff
@@ -0,0 +1,273 @@
...
+    |> Stream.map(fn line ->
+      [token | entities] = String.split(line, @def_split)
+      if(
+      !is_nil(token) &&
+      token !== "" &&
+      String.length(token) !== 1 &&
+      line !== "\n" &&
+      token !== @start &&
+      token !== @ending &&
+      token not in @incorrect)
```
Comment (jonatanklosko) [inline examples/classifcation/text_classification.exs:181]:
```suggestion
      if token not in ["", @start, @ending] and token not in @incorrect and String.length(token) !== 1
```

I don't think `token` can ever be `nil` :)

File: examples/classifcation/text_classification.exs [hunk 14]
```diff
@@ -0,0 +1,273 @@
...
+      do
+
+        [ r | rest ] = entities |> List.last |> String.split("\n")
```
Comment (jonatanklosko) [inline examples/classifcation/text_classification.exs:184]: We are processing lines from the stream, right? Why do we need to split on newline here?

File: examples/classifcation/text_classification.exs [hunk 15]
```diff
@@ -0,0 +1,273 @@
...
+
+        {String.downcase(token), r}
+      else
+        nil
+      end
+    end)
+    |> Stream.reject(fn x -> is_nil(x) end)
```
Comment (jonatanklosko) [inline examples/classifcation/text_classification.exs:191]:
```suggestion
    |> Stream.reject(&is_nil/1)
```

File: examples/classifcation/text_classification.exs (hunk 13 above)
Comment (fantypants) [reply examples/classifcation/text_classification.exs:181]: Just validated this & also changed some things based on suggestion. The reason for it there was the weird early handling of the row:
`"WORD LABEL LABEL LABEL\n"` when I adjusted the preprocessing as you suggested below it was much more obvious.

File: examples/classifcation/text_classification.exs (hunk 14 above)
Comment (fantypants) [reply examples/classifcation/text_classification.exs:184]: ^Above comment relates to this, all fixed and redone =)

**Comment by fantypants** [comment]
> 

@seanmor5 I've finished the changes and pushed to my main branch, how do I update the PR with the additions?

**Comment by jdewar** [comment]
filename has one too many s's, and directory has one too few i's

`examples/classifcation/text_classsification.exs` -> `examples/classification/text_classification.exs`

File: examples/classifcation/text_classsification.exs [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+  def train(model, data) do
+    model
+    |> Axon.Loop.trainer(:kl_divergence, Axon.Optimizers.adam(0.001)) # Can use :categorical_cross_entropy
+    |> Axon.Loop.metric(:accuracy, "Accuracy")
+    |> Axon.Loop.run(data, %{}, epochs: 20, iterations: 300)
+  end
+
+  defp kl_divergence(y_true, y_pred) do
+    Axon.Losses.kl_divergence(y_true, y_pred, reduction: :mean)
+  end
```
Comment (seanmor5) [inline examples/classifcation/text_classsification.exs:45]:
```suggestion
```

I don't think you need this anymore, right?

File: examples/classifcation/text_classsification.exs [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+    train_labels =
+      word_labels
+      |> Enum.map(fn {word, label} ->
+          encode(label_to_idx, label)
+      end)
+      |> Enum.chunk_every(@sequence_length, 1, :discard)
+      |> Enum.drop(-1)
+      |> Nx.tensor
+      |> Nx.reshape({:auto, @sequence_length, 1})
+      |> Nx.to_batched_list(@batch_size)
```
Comment (seanmor5) [inline examples/classifcation/text_classsification.exs:83]:
```suggestion
      |> Nx.to_batched(@batch_size)
```

File: examples/classifcation/text_classsification.exs [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+      word_labels
+      |> Enum.map(fn {word, label} ->
+        encode(word_to_idx, word)
+      end)
+      |> Enum.chunk_every(@sequence_length, 1, :discard)
+      |> Enum.drop(-1)
+      |> Nx.tensor
+      |> Nx.divide(wcount)
+      |> Nx.reshape({:auto, @sequence_length})
+      |> Nx.to_batched_list(@batch_size)
```
Comment (seanmor5) [inline examples/classifcation/text_classsification.exs:71]:
```suggestion
      |> Nx.to_batched(@batch_size)
```

**Comment by seanmor5** [comment]
@fantypants Hey! Just checking up on the status of this PR? Do you need help or another review?

**Comment by fantypants** [comment]
> @fantypants Hey! Just checking up on the status of this PR? Do you need help or another review?

Hey @seanmor5, Apologies for the delay, I do believe it would be good for a review soon; however, I'll take a look over it tomorrow and fix another highlighted issue with the filename. I also have to go through the changes of late to make sure it's all still compatible. I'll give you a shout on here when it's ready (1-2 Days).
